### PR TITLE
Signup API Endpoint and mongoose Integration

### PR DESCRIPTION
**Pull request by franceschiniandrea**
Added mongoose and developed API endpoint for signup at `api/auth/signup`

**Sign Up Logic**
1. Connect to Mongoose
2. Pull body from Request
3. Check if password is present and hash password
4. Create a User from the mongoose Schema
5. Save the user and handle any validation error

**Other features included to make Signup work**
- Database utility to connect and disconnect from mongoose (lib/database.ts)
  - Logic to avoid having more than 1 connection at a time
- Mongoose
  - User Model with validation (models/userModel.ts) 
- Tests
  - Endpoint tests
  - Mongoose model validation tests
  - Changed `jest.config` to make tests work with mongoose
- Error handler (lib/handleError.ts)
- Password Hasher using `bcrypt` (lib/hashPassword.ts)

### CODE REVIEW

File: @types/global.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+		promise: Promise<typeof mongooseInstance> | null
+	}
+
+	/**
+	 * `global.mongoose` is used to maintain only *ONE*
+	 * active connection to mongoose at a time, by caching it in a global variable.
+	 *
+	 * This prevents connections from growing exponentially during API Routes usage
+	 */
+	var mongoose: MongooseGlobalVar
```
Comment (moody2times) [inline @types/global.d.ts:18]: Any reason why you used `var` here instead of `let`?

**Review by moody2times — CHANGES_REQUESTED** [review]
Just one tiny little snag and then I will merge this.

File: @types/global.d.ts (hunk 1 above)
Comment (franceschiniandrea) [reply @types/global.d.ts:18]: Actually, I tried using `let` but the type won't work that way. 
Here is the video:

https://user-images.githubusercontent.com/65458493/184605493-bd2a1f4b-9518-45b3-9694-6f8e5362ff56.mov

Comment (moody2times) [reply @types/global.d.ts:18]: Ok then.

**Review by moody2times — APPROVED** [review]
Merging this in

**Pull request merged**